### PR TITLE
Reduce intermediate collections in NameRes

### PR DESCRIPTION
**Pull request by forki**

### CODE REVIEW

**Comment by dsyme** [comment]
Some bits look ok but some looks smelly. The parts that look ok are the parts that deal with empty sets of vrefs, (though they might have been picked up by earlier shotcuts?)

Smelly bits: for example https://github.com/Microsoft/visualfsharp/pull/4437/files#diff-dbb9e59b8ac1e1e04e1b00282a91ec97R583 is callin ``Seq.isEmpty``.  That is a more expensive operation that ``Array.isEmpty`` or ``arr.Length = 0`` - Seq.isEmpty creates an iterator and moves one step.  That's itself an allocating operation. Also, ``Seq.fold`` is roughly equivalent to creating an intermediate array and then folding over that - again at least an iterator is created.  So just converting to an array and going from there can be simpler.

In short let's not use sequences in this code please.  Also the laziness of sequence makes me nervous when hunting bugs in the code.

**Comment by dsyme** [comment]
To explain the BulkAdd stuff - my belief is that the really big win in this code for long-lived memory usage (not short-lived allocations) is when processing a long series of open declarations

    open System.This
    ...
    open System.That
    ...
    user code
    ...
```
My gut feeling is that the functional maps we use should have two "layers" - a layer of long-lived data in a "base layer", and a layer of incrementally added data in the "top layer".  The base layer was going to use an immutable dictionary.  

The BulkAdd code was part of an attempt to do this. We would use BulkAdd during "open" operations and then pushing everything that results into the base layer.  But the code has rotted and can be removed. And I'm not even sure my intuitions are correct here.  We might get more mileage by just switching to use a 3-or-4-way switching hash-trie.  Right now the maps must do a lot of string comparisons on the way down each lookup.

Basically it feels like the current maps we use are ok for small maps, but aren't particularly optimized to the case where we have 10,000 types indexed by name, and then add a coupe of extra type definitions on top of these.

**Comment by forki** [comment]
yeah tried to replace the bulkadd where it's essentially doing the same stuff in both cases. but didn't really turn out to be helping much
